### PR TITLE
Ensure intended results for optimizer rule tests

### PR DESCRIPTION
**Pull request by dey4ss**
During recent work, we considerably often found the following pattern in optimizer rule tests:
```c++
const auto input_lqp = ...;
const auto result_lqp = apply_rule(rule, input_lqp);
const auto expected_lqp = input_lqp->deep_copy();
EXPECT_LQP_EQ(result_lqp, expected_lqp);
```

With the current implementation of `StrategyBaseTest::apply_rule(...)`, this approach has multiple flaws:
1. `apply_rule(...)` created a `LogicalPlanRootNode`, set the given LQP as input, applied the rule to the root node, and returned the root node's input. In many cases, THIS WAS JUST THE POINTER TO THE GIVEN LQP.
2. The expected LQP was copied from the input LQP after applying the rule. With the consideration above, this means we compare the optimized LQP with a copy of itself. Obviously, this is always correct (but not what we want).
3. Even if the expected LQP is not a direct copy of the input/optimized LQP, it often shares nodes with it. Thus, both LQPs are identical for the shared nodes, which is not always what is intended.

In detail:
```c++
const auto input_lqp = ...;
const auto result_lqp = apply_rule(rule, input_lqp);  // input_lqp and result_lqp are pointers to the same
                                                      // (optimized) LQP whenever the rule does not replace
                                                      // the top node by another one.
const auto expected_lqp = input_lqp->deep_copy();     // This is now a copy of the optimized plan.
EXPECT_LQP_EQ(result_lqp, expected_lqp);              // Trivially correct.
```

The issues arose mostly from the fact that `apply_rule(...)` returned `std::shared_ptr<AbstractLQPNode>`. It was not clear that it did not optimize a copy of the passed plan but directly optimized the given LQP. In other test cases, however, it was correctly assumed that the input LQP changes.

Thus, this PR changes the method's signature to be `void`. Now, it is clear that the passed LQP is optimized by the passed rule. In order to do so, we had to change the interface of `StrategyBaseTest` to ease assigning different node types to the same variables. The relevant issues are now described in the `StrategyBaseTest` header file. Doing so, we were forced to revisit all calls of the method. Indeed, some test cases did not actually check what they were intended to to. This is now fixed.

Besides, we applied some style guidelines.

Many of the changes are trivial to review, do not be afraid of the large diff.

### CODE REVIEW

File: src/test/lib/optimizer/strategy/in_expression_rewrite_rule_test.cpp [hunk 1]
```diff
@@ -238,99 +245,99 @@ TEST_F(InExpressionRewriteRuleTest, JoinStrategy) {
...
     const auto right_col = lqp_column_(static_table_node, ColumnID{0});
     const auto expected_lqp = JoinNode::make(JoinMode::Semi, equals_(col_a, right_col), node, static_table_node);
 
-    EXPECT_LQP_EQ(result_lqp, expected_lqp);
-    EXPECT_TABLE_EQ_UNORDERED(static_cast<StaticTableNode&>(*result_lqp->right_input()).table, table);
+    EXPECT_LQP_EQ(_lqp, expected_lqp);
+    EXPECT_TABLE_EQ_UNORDERED(static_cast<StaticTableNode&>(*_lqp->right_input()).table, table);
 
-    EXPECT_NEAR(cardinality_estimator.estimate_cardinality(result_lqp), 1000.f / 200 * 100, 10);
+    EXPECT_NEAR(cardinality_estimator.estimate_cardinality(_lqp), 1000.f / 200 * 100, 10);
```
Comment (Bouncner) [inline src/test/lib/optimizer/strategy/in_expression_rewrite_rule_test.cpp:331]: Just curious. In the FK PR, we used an epsilon and did the comparison manually. Right?

**Review by Bouncner — COMMENTED** [review]
A few questions.

File: src/test/lib/optimizer/strategy/stored_table_column_alignment_rule_test.cpp [hunk 2]
```diff
@@ -82,27 +82,27 @@ TEST_F(StoredTableColumnAlignmentRuleTest, DifferentTableEqualChunksDifferentCol
...
-  auto projection_subquery = ProjectionNode::make(expression_vector(column_c), stn_subquery);
-  auto subquery = lqp_subquery_(projection_subquery);
+  // Prepare root & subquery LQP.
+  const auto stn_subquery = std::static_pointer_cast<StoredTableNode>(_stored_table_node_left->deep_copy());
+  const auto column_c = lqp_column_(stn_subquery, ColumnID{2});
+  const auto projection_subquery = ProjectionNode::make(expression_vector(column_c), stn_subquery);
+  const auto subquery = lqp_subquery_(projection_subquery);
 
-  auto projection_root = ProjectionNode::make(expression_vector(subquery), _union_node);
+  _lqp = ProjectionNode::make(expression_vector(subquery), _union_node);
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/lib/optimizer/strategy/stored_table_column_alignment_rule_test.cpp:90]: Ah ... needed to fix a few issues with the merge and I guess I understand now.

File: src/test/lib/optimizer/strategy/column_pruning_rule_test.cpp [hunk 3]
```diff
@@ -53,64 +53,55 @@ class ColumnPruningRuleTest : public StrategyBaseTest {
...
     PredicateNode::make(greater_than_(5, c),
       JoinNode::make(JoinMode::Inner, greater_than_(v, a),
         node_abc,
-        SortNode::make(expression_vector(w), std::vector<SortMode>{SortMode::Ascending},  // NOLINT
-          node_uvw))));
-
-  // Create deep copy so we can set pruned ColumnIDs on node_abc below without manipulating the input LQP.
-  lqp = lqp->deep_copy();
+        SortNode::make(expression_vector(w), std::vector<SortMode>{SortMode::Ascending},
+          node_uvw))))->deep_copy();
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/lib/optimizer/strategy/column_pruning_rule_test.cpp:62]: Got it.

What would happen when nodes are shared? Can we somehow check when users should have used `deep_copy()`?

File: src/test/lib/optimizer/strategy/stored_table_column_alignment_rule_test.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/lib/optimizer/strategy/stored_table_column_alignment_rule_test.cpp:90]: But what I still don't understand is the error when not using a abstractlqpnode pointer:
`error: non-const lvalue reference to type 'shared_ptr<AbstractLQPNode>' cannot bind to a value of unrelated type 'shared_ptr<hyrise::PredicateNode>'`

Why are they considers not to be related? PredicateNode inherits from AbstractLQPNode

[the remark this answers is not in the document]
Comment (dey4ss) [reply src/test/lib/optimizer/strategy/stored_table_column_alignment_rule_test.cpp:90]: Would need more details on this example, but looks like you tried to reassign a reference?

Your example should work, but we have a problem with doing the following, where `_rule` is a `PredicateSplitUpRule`:
```c++
auto lqp = PredicateNode::make(or_(equals_(a, 3), equals(b, 4)), table);
_apply_rule(_rule, lqp); // _rule creates a UnionNode with two PredicateNodes as input.
                         // ERROR: incompatible types! 
```
so we must either retain from auto-to-stick for `lqp`, make a deep copy, or cast to an `AbstractLQPNode` pointer. Guess that's the reason that `apply_rule` returned the new root.

As you see, I fiddled around the first option by adding the member variable. A deep copy is also not always wanted when we check that things in the original LQP changed, e.g., columns were pruned.

File: src/test/lib/optimizer/strategy/chunk_pruning_rule_test.cpp [hunk 4]
```diff
@@ -548,13 +524,12 @@ TEST_F(ChunkPruningRuleTest, DoNotSetPrunableSubqueryScansComplicatedPredicate)
...
 
-  const auto input_lqp =
+  _lqp =
   PredicateNode::make(in_(stored_table_node_1_col_a, lqp_subquery_(subquery)),
     stored_table_node_1);
   // clang-format on
 
-  const auto actual_lqp = StrategyBaseTest::apply_rule(_rule, input_lqp);
-  EXPECT_EQ(actual_lqp, input_lqp);
+  _apply_rule(_rule, _lqp);
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/test/lib/optimizer/strategy/chunk_pruning_rule_test.cpp:531]: We just check here that the pruned columns of nodes in the input LQP are set (which were unset before). See my comment above.

File: src/test/lib/optimizer/strategy/index_scan_rule_test.cpp [hunk 5]
```diff
@@ -75,23 +68,23 @@ TEST_F(IndexScanRuleTest, NoIndexScanWithIndexOnOtherColumn) {
...
 
-  auto predicate_node_0 = PredicateNode::make(equals_(a, 10));
-  predicate_node_0->set_left_input(stored_table_node);
+  _lqp = PredicateNode::make(equals_(a, 10));
+  _lqp->set_left_input(stored_table_node);
 
-  EXPECT_EQ(predicate_node_0->scan_type, ScanType::TableScan);
-  StrategyBaseTest::apply_rule(rule, predicate_node_0);
-  EXPECT_EQ(predicate_node_0->scan_type, ScanType::TableScan);
+  EXPECT_EQ(static_cast<PredicateNode&>(*_lqp).scan_type, ScanType::TableScan);
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/test/lib/optimizer/strategy/index_scan_rule_test.cpp:80]: Basically just consistency here. Also thought about `apply_rule()` not to take any parameters at all and directly use `_rule` and `_lqp` members of `StrategyBaseTest`.

File: src/test/lib/optimizer/strategy/in_expression_rewrite_rule_test.cpp [hunk 6]
```diff
@@ -52,7 +54,8 @@ class InExpressionRewriteRuleTest : public StrategyBaseTest {
  public:
   // Can't use EXPECT_LQP_EQ for disjunction rewrites for multiple elements, because ExpressionUnorderedSet produces
   // a non-deterministic order of predicates
-  bool check_disjunction(std::shared_ptr<AbstractLQPNode> result_lqp, std::vector<int> expected_values) {
+  bool check_disjunction(std::shared_ptr<AbstractLQPNode> lqp, std::shared_ptr<AbstractExpression> expected_column,
+                         std::vector<int> expected_values) {
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/test/lib/optimizer/strategy/in_expression_rewrite_rule_test.cpp:57]: gotta look into it, sounds valid.

File: src/test/lib/optimizer/strategy/join_to_predicate_rewrite_rule_test.cpp [hunk 7]
```diff
@@ -63,30 +64,29 @@ TEST_P(JoinToPredicateRewriteRuleJoinModeTest, PerformRewrite) {
...
-  const auto lqp =
+  _lqp =
   ProjectionNode::make(expression_vector(b),
     join_node);
 
   const auto subquery = ProjectionNode::make(expression_vector(u),
     PredicateNode::make(equals_(v, 0), node_b));
 
-  auto expected_lqp =
+  auto expected_lqp = std::static_pointer_cast<AbstractLQPNode>(
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/test/lib/optimizer/strategy/join_to_predicate_rewrite_rule_test.cpp:75]: Honestly, have to look again into what I did 10 months ago,

File: src/test/lib/optimizer/strategy/in_expression_rewrite_rule_test.cpp (hunk 1 above)
Comment (dey4ss) [reply src/test/lib/optimizer/strategy/in_expression_rewrite_rule_test.cpp:330]: `EXPECT_NEAR` expects an absolute value. In the FK PR, the cardinality depends on the SF and so should the epsilon. But I guess `EXPECT_NEAR(actual, expected, expected * epsilon);` should be fine there.

File: src/test/lib/optimizer/strategy/chunk_pruning_rule_test.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/lib/optimizer/strategy/chunk_pruning_rule_test.cpp:531]: What I meant was: why remove the EXPECT_EQ? (or expect_lqp_eq).

And we mix expect and assert quite a lot here.

[the remark this answers is not in the document]
Comment (dey4ss) [reply src/test/lib/optimizer/strategy/chunk_pruning_rule_test.cpp:531]: Did not look into `ASSERT` vs `EXPECT` too much. Whenever I add it, I try to use `ASSERT` when the following code would segfault or have out-of-bounds accesses.
The thing is that the input `_lqp` IS the one that is optimized. What should we compare it with here?

[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/lib/optimizer/strategy/chunk_pruning_rule_test.cpp:531]: Got it.

File: src/test/lib/optimizer/strategy/join_to_predicate_rewrite_rule_test.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/test/lib/optimizer/strategy/join_to_predicate_rewrite_rule_test.cpp:75]: We do not rewrite the plan if it is an inner or semi join. Thus, we replace the expected plan with a deep copy of the input for other join nodes. So wither we cast here or when assigning the copy. Fine with both.

**Review by Bouncner — APPROVED** [review]
LGTM

**Pull request merged**